File: algorithms/valid_palindrome.js
/**
 * @param {string} s
 * @return {boolean}
 */

function isAlpha(c) {
  c = c.toLowerCase().charCodeAt(0);
  return (48 <= c && c <= 57) || (97 <= c && c <= 122);
}

var isPalindrome = function(s) {
    let [l, r] = [0, s.length - 1];
    while (l <= r) {
        if (!isAlpha(s[l])) l++;
        else if (!isAlpha(s[r])) r--;
        else if (s[l].toLowerCase() !== s[r].toLowerCase()) return false;
        else {
          l++;
          r--;
        }
    }
    return true;
};

File: java/ValidPalindrome.java
class ValidPalindrome {
    public boolean isPalindrome(String s) {
        char[] chars = s.toCharArray();

        int left = 0, right = chars.length - 1;
        while (left < right) {
            if (!Character.isLetterOrDigit(chars[left])) {
                left++;
            } else if (!Character.isLetterOrDigit(chars[right])) {
                right--;
            } else if (Character.toLowerCase(chars[left]) != Character.toLowerCase(chars[right])) {
                return false;
            } else {
                left++;
                right--;
            }
        }

        return true;
    }
}